### ISSUE TITLE
Behaviour of add methods with null value

### ISSUE THREAD

**Issue by jmini** [eclipse/microprofile-open-api#323]
When a map is used, the model interfaces have `addXxxx(String, T)` methods.

The Javadoc is not clear about `null` beeing accepted or not as value of type T

I think that calling an add method with a `null` value (e.g. `addXxxx(“key”, null)`) is equivalent to calling `removeXxxx(“key”)`. 

---
The test in `ModelConstructionTest` do not enforce this. We could add something like:

```java
Components components = OASFactory.createComponents();

components.addSchema(
        "MyObject", OASFactory.createSchema()
                .type(SchemaType.STRING)
);
// check on components.getSchemas() => it should only contains "MyObject"

components.addSchema("MyObject", null);
// check on components.getSchemas() => it should be empty
```
---

List:
```
Components#addSchema(String, Schema)
Components#addResponse(String, APIResponse)
Components#addParameter(String, Parameter)
Components#addExample(String, Example)
Components#addRequestBody(String, RequestBody)
Components#addHeader(String, Header)
Components#addSecurityScheme(String, SecurityScheme)
Components#addLink(String, Link)
Components#addCallback(String, Callback)

Operation#addCallback(String, Callback)

Paths#addPathItem(String, PathItem)

Callback#addPathItem(String, PathItem)

Header#addExample(String, Example)

Link#addParameter(String, Object)

Content#addMediaType(String, MediaType)

Discriminator#addMapping(String, String)

Encoding#addHeader(String, Header)

MediaType#addExample(String, Example)
MediaType#addEncoding(String, Encoding)

Schema#addProperty(String, Schema)

Parameter#addExample(String, Example)

APIResponse#addHeader(String, Header)
APIResponse#addLink(String, Link)

APIResponses#addAPIResponse(String, APIResponse)

Scopes#addScope(String, String)

SecurityRequirement#addScheme(String, List)

ServerVariables#addServerVariable(String, ServerVariable)
```

---
Example found in a vendor implementation:
```java
    /**
     * @see org.eclipse.microprofile.openapi.models.responses.APIResponses#addAPIResponse(java.lang.String, org.eclipse.microprofile.openapi.models.responses.APIResponse)
     */
    @Override
    public APIResponses addAPIResponse(String name, APIResponse apiResponse) {
        if (apiResponse == null) {
            this.removeAPIResponse(name);
        } else {
            this.put(name, apiResponse);
        }
        return this;
    }
```

**Comment by jmini** [comment]
Alternative option:

Indicate in the Javadoc that for the add methods, the value (second parameter) "should not" or "must not" be null.
This could be enforced in the TCK or not.

**Comment by arthurdm** [comment]
hey @jmini - I prefer using the `removeX` methods to remove items, so I am in favour of your alternative option to specify in the javadoc that the `value` must not be null.

**Comment by EricWittmann** [comment]
I agree with @arthurdm on this - I don't think the `addX` methods should be able to remove items.  So my vote goes to the alternative option - and I think the TCK should enforce it.

**Comment by jmini** [comment]
Improving the Javadoc is OK.

What is the expected enforcement for the TCK?

1. `addXxxx("key", null)` **CAN** throw an exception (like `IllegalArgumentException`, but I think the TCK should not enforce the type of Exception)
2. after `addXxxx("key", null)`, calling `containsKey("key")` on the corresponding Map **MUST** return false after.

For 1. it could also be a MUST, but I think CAN is better for the vendors.

**Comment by EricWittmann** [comment]
I agre - I think (2) is fine for the TCK.  There's no need to dictate exception throwing.

**Comment by arthurdm** [comment]
We probably just need to be careful on the wording of (2) to say that passing `null` values will be ignored, instead of saying that it will return `false` - as in some cases it will return `false`, but not all. 

Consider this scenario:

```
addX("myKey", "myValue")
addX("myKey", null)
```
After this executes, should `containsKey("myKey")` return `true` or `false`?  I think it should return `true`, if `null` values are ignored.

**Comment by jmini** [comment]
PR for 2.0: #325

**Comment by jmini** [comment]
Findings by @EricWittmann:

> 1.
> https://github.com/eclipse/microprofile-open-api/blob/bd7706d3bacfc15e8253978b75348f54df1b232f/tck/src/main/java/org/eclipse/microprofile/openapi/tck/ModelConstructionTest.java#L823
> I don't think description is required for a scope in a flow.
> It would just add a scope without a description
> From the spec: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#oauth-flow-object
> I think we should assume unless the spec says it's mandatory, then it's optional.

And:

> 2.
> I've got one more - the `SecurityRequirement` `addScheme` methods should allow nulls

**Comment by jmini** [comment]
All PRs are merged:

* `master` branch
  * in release `1.1`: #324 
  * in release `1.1.1` additional fixes #329
* And merged into `2.0` branch with #325

Closing this issue now.
